File: TypeScript/Algorithm/Algorithm I/876. Middle of the Linked List.ts
/**
 * Definition for singly-linked list.
 * @class ListNode
 * @property {number} val - The value of the node.
 * @property {ListNode | null} next - The next node in the linked list.
 * @constructor Create a ListNode.
 * @param {number} [val] - The value of the node.
 * @param {ListNode | null} [next] - The next node in the linked list.
 *
 * class ListNode {
 *     val: number;
 *     next: ListNode | null;
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val === undefined ? 0 : val);
 *         this.next = (next === undefined ? null : next);
 *     }
 * }
 */

/**
 * Finds the middle node of a linked list.
 * @param {ListNode | null} head - The head node of the linked list.
 * @returns {ListNode | null} The middle node of the linked list.
 */
function middleNode(head: ListNode | null): ListNode | null {
  let fastTraverse: ListNode | null = head;
  let slowTraverse: ListNode | null = head;

  while (slowTraverse && fastTraverse && fastTraverse.next) {
    fastTraverse = fastTraverse.next.next;
    slowTraverse = slowTraverse.next;
  }

  return slowTraverse;
}

File: Java/Algorithm/Algorithm I/876. Middle of the Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

  /**
   * Finds the middle node of a singly-linked list.
   *
   * @param head The head of the linked list.
   * @return The middle node of the linked list.
   */
  public ListNode middleNode(ListNode head) {
    ListNode fastList = head;
    ListNode slowList = head;

    while (fastList != null && fastList.next != null) {
      fastList = fastList.next.next;
      slowList = slowList.next;
    }

    return slowList;
  }
}
